host/xtest/asym_perf.c

```diff
@@ -0,0 +1,962 @@
+		break;
+	}
+
+	prepare_obj(width_bits, main_algo);
+
+	res = pack_attrs(params, param_count, &buf, &blen);
+	CHECK(res, "pack_attrs", goto out;);
+
+	if (mode == MODE_GENKEYPAIR) {
+		if (blen)
+			prepare_attrs(buf, blen);
+
+		cmd = TA_CRYPTO_PERF_CMD_ASYM_CIPHER_PROCESS_GEN_KEYPAIR;
+		op.paramTypes = TEEC_PARAM_TYPES(TEEC_VALUE_INPUT,
+						 TEEC_NONE, TEEC_NONE,
+						 TEEC_NONE);
+		op.params[0].value.a = width_bits;
+		op.params[0].value.b = l;
+	} else {
+		prepare_keypair(width_bits, buf, blen, mode, algo, main_algo);
+
+		if (mode == MODE_SIGN || mode == MODE_VERIFY)
+			prepare_hash(size, main_algo);
+
+		if (main_algo == RSA && algo >= RSASSA_PKCS1_PSS_MGF1_SHA1) {
+			params[0].attributeID = TEE_ATTR_RSA_PSS_SALT_LENGTH;
+			params[0].content.value.a = salt_len;
+			params[0].content.value.b = 0;
+			param_count = 1;
+			res = pack_attrs(params, param_count, &buf, &blen);
+			CHECK(res, "pack_attrs", goto out;);
+		}
+
+		if (blen)
+			prepare_attrs(buf, blen);
+
+		cmd = TA_CRYPTO_PERF_CMD_ASYM_CIPHER_PROCESS;
+		op.paramTypes = TEEC_PARAM_TYPES(TEEC_VALUE_INPUT,
+						 TEEC_MEMREF_PARTIAL_INPUT,
+						 TEEC_MEMREF_PARTIAL_OUTPUT,
+						 TEEC_NONE);
+
+		op.params[0].value.a = l;
+		op.params[0].value.b = mode;
+
+		switch (mode) {
+		case MODE_ENCRYPT:
+			op.params[1].memref.parent = &in_shm;
+			op.params[1].memref.size = size;
+			op.params[2].memref.parent = &out_shm;
+			op.params[2].memref.size = out_shm.size;
+			break;
+		case MODE_SIGN:
+			op.params[1].memref.parent = &hash_shm;
+			op.params[1].memref.size = hash_shm.size;
+			op.params[2].memref.parent = &out_shm;
+			op.params[2].memref.size = out_shm.size;
+			break;
+		case MODE_DECRYPT:
+			prepare_enc_sign(size, mode, is_random, buf, blen);
+			op.params[1].memref.parent = &out_shm;
+			op.params[1].memref.size = out_shm.size;
+			op.params[2].memref.parent = &in_shm;
+			op.params[2].memref.size = size;
+			break;
+		case MODE_VERIFY:
+			prepare_enc_sign(size, mode, is_random, buf, blen);
+			op.params[1].memref.parent = &hash_shm;
+			op.params[1].memref.size = hash_shm.size;
+			op.params[2].memref.parent = &out_shm;
+			op.params[2].memref.size = out_shm.size;
+			break;
+		default:
+			fprintf(stderr, "Unexpected mode value\n");
+			exit(1);
+		}
+	}
+
+	if (warmup)
+		do_warmup(warmup);
+
+	while (n-- > 0) {
+		if (mode == MODE_ENCRYPT && main_algo == RSA) {
+			/* Make sure the N > M */
+			((unsigned char *)(in_shm.buffer))[0] = 0x00;
+			/* Avoid the problem that the last encryption result is
+			 * shorter than the plaintext.
+			 */
+			op.params[3].memref.size = out_shm.size;
+		}
+
+		get_current_time(&t0);
+
+		res = TEEC_InvokeCommand(&sess, cmd, &op, &ret_origin);
+
+		check_res(res, "TEEC_InvokeCommand()", &ret_origin);
+
+		get_current_time(&t1);
+
+		update_stats(&stats, timespec_diff_ns(&t0, &t1));
+		if (n % (n0 / 10) == 0)
+			vverbose("#");
+	}
+
+	vverbose("\n");
+	sd = stddev(&stats);
+	printf("%.3f ops/s\n", l / stats.m);
```

Comment:
I would like to come back on a previous comment I made about adding `ops/s` measurement.
It would make more sense to have `ms/op` instead: https://www.cryptopp.com/benchmarks.html.
`printf("%.3f ms/op\n", stats.m / 1000000 / loops);`